lms/services/canvas_api/client.py

```diff
@@ -200,23 +201,45 @@ def post_load(self, data, **_kwargs):  # pylint:disable=no-self-use
                 for enrollment in data["enrollments"]
             ]
 
-    def list_files(self, course_id):
+    def list_files(self, course_id, request=None):
         """
         Return the list of files for the given `course_id`.
 
         :param course_id: the Canvas course_id of the course to look in
+        :param request: Pyramid request object used to store file details if
+            present
         :rtype: list(dict)
         """
         # For documentation of this request see:
         # https://canvas.instructure.com/doc/api/files.html#method.files.api_index
 
-        return self._client.send(
+        files = self._client.send(
             "GET",
             f"courses/{course_id}/files",
             params={"content_types[]": "application/pdf"},
             schema=self._ListFilesSchema,
         )
 
+        if request:
+            # Notify that we've found some files
+            request.registry.notify(
+                FilesDiscoveredEvent(
+                    request=request,
+                    values=[
```

Comment:
Would it make sense that FilesDiscoveredEvent takes the full files objects instead of just some `values`? Then it's up to the event what to do with them

Reply:
It would ensure they have the right shape, but we'd need to marshall them back down to dicts again I think. So it's nice from the point of view of using the model code to enforce the schema, but it might add extra code.

I'll give it a shot and see if it works. Maybe SQLAlchemy will just accept it?

Reply:
I've given this a shot, but the insert function expects plain dicts, so we have to convert them back again. I tried using a generic method, but as we don't set most of the columns it means weeding out `None`'s. This also means distinguishing between `None`'s that are there by default and `None`'s that we might be trying to set... all of which is to say... maybe just not do this?